app/lib/primer/classify/utilities.rb

```diff
@@ -0,0 +1,114 @@
+# frozen_string_literal: true
+
+# :nodoc:
+module Primer
+  class Classify
+    # Handler for PrimerCSS utility classes loaded from utilities.rake
+    class Utilities
+      class << self
+        def classname(key, val, breakpoint = "")
+          if (valid = validate(key, val, breakpoint))
+            valid
+          else
+            # Get selector
+            Primer::Classify::UTILITIES[key][val][Primer::Classify::BREAKPOINTS.index(breakpoint)]
+          end
+        end
+
+        # Does the Utilitiy class support the given key
+        #
+        # returns Boolean
+        def supported_key?(key)
+          Primer::Classify::UTILITIES[key].present?
+        end
+
+        # Does the Utilitiy class support the given key and value
+        #
+        # returns Boolean
+        def supported_value?(key, val)
+          supported_key?(key) && Primer::Classify::UTILITIES[key][val].present?
+        end
+
+        # Is the key and value responsive
+        #
+        # returns Boolean
+        def responsive?(key, val)
+          supported_value?(key, val) && Primer::Classify::UTILITIES[key][val].count > 1
+        end
+
+        # Get the options for the given key
+        #
+        # returns Array or nil if key not supported
+        def mappings(key)
+          return unless supported_key?(key)
+
+          Primer::Classify::UTILITIES[key].keys
+        end
+
+        # Extract hash from classes ie. "mr-1 mb-2 foo" => { mr: 1, mb: 2, classes: "foo" }
+        def classes_to_hash(classes)
+          obj = {}
+          classes = classes.split(" ")
+          classes.reject! do |c|
+            key, value, index = find_selector(c)
+            next false if key.nil?
+
+            # Create array if nil
+            obj[key] = Array.new(5, nil) if obj[key].nil?
+            obj[key][index] = value
+            next true
+          end
+
+          # Transform responsive arrays into arrays without trailing nil
+          obj.transform_values! do |v|
+            v = v.reverse.drop_while(&:nil?).reverse
+            if v.count == 1
+              v.first
+            else
+              v
+            end
+          end
+
+          obj[:classes] = classes.join(" ") if classes.any?
+          obj
+        end
+
+        private
+
+        def find_selector(selector)
+          Primer::Classify::UTILITIES.each do |k, v|
+            v.each do |vk, vv|
+              next unless vv.include?(selector)
+
+              return [k, vk, vv.index(selector)]
+            end
+          end
```

Comment:
If I understand correctly, we are traversing the whole YAML file looking for the class and then returning the keys. Performance could be an issue here (not sure if we care about it though).

Reply:
Yeah I suspect it would be slow, but it's not something we'll do outside of ci in prod?

Reply:
Maybe I can guard that method with a Rails.env.production? check to return a {} just so we don't accidentally use it in production.